sockeye/utils.py

```diff
@@ -264,6 +264,124 @@ def average_tensors(tensors: List[pt.Tensor]) -> pt.Tensor:
     return sum(tensors) / len(tensors)  # type: ignore
 
 
+def one_hot_encoding_from_prefix(prefix: pt.Tensor, vocab_size: int) -> pt.Tensor:
+    """
+    Create an one hot encoding from output vocab size of prefix
+
+    :param prefix: Target prefix ids. Shape (batch size, 1).
+    :param vocab_size: vocabulary size
+    :return prefix_in_one_hot (batch size, 1, output_vocab_size)
+
+    """
+    prefix_in_one_hot = pt.nn.functional.one_hot(prefix.to(pt.int64), num_classes=vocab_size).to(prefix.device)
+    # all hots are assigned to 1 if a prefix id is 0 (i.e. no prefix)
+    prefix_in_one_hot.masked_fill_(prefix.unsqueeze(-1) == 0, 1)
```

Comment:
I don't think I understand this line. Isn't the one-hot from line 276 already only 0s and 1s? I assume `prefix` is `(batch size, 1)` and looks like this: 
```
prefix = pt.tensor([1,2,3]).unsqueeze(1)
```
What does this line do then?

Reply:
Thanks for this. I definitely should add more explanation regarding this. So now I added a detailed explanation there. Pls let me know if it is still not that clear to you.

```
    # In the same batch during inference, it is possible that some translations have target prefix
    # while others do not have. It is also possible that translation may have a target prefix with
    # different length to others. Thus prefix ids may include a full zero vector if a translation
    # in the batch does not have prefix, or include a vector padding with zeros on the right if some
    # translations are with shorter prefix. An example of prefix ids reflecting length differences \
    # is as follows:
    #
    # [1, 2, 3]
    # [1, 2, 0]
    # [0, 0, 0]
    #
    # Here, the first sentence has a prefix of length 3, the second one has a prefix of length 1 \
    # and the last one does not have prefix.
    #
    # At any timestep, some target prefix ids could be zero. If a prefix id is zero for a translation \
    # at a timestep, all hots in the vocab are assigned to 1 (instead of only one hot is assigned to 1). \
    # This makes sure there is no constraint on selecting any specific target token for the translation
    # in that case.

    prefix_in_one_hot.masked_fill_(prefix.unsqueeze(-1) == 0, 1)
```

Reply:
oh I see, the semantics of `0` / `C.PAD_ID` in the `target_prefix` weren't clear to me. `0` in the `target_prefix` means 'no constraint'